### PR TITLE
os/bluestore/NVMEDevice: add the spdk core mask check

### PR DESCRIPTION
**Pull request by optimistyzy**
This patch adds the spdk core mask check and also
set the master core for starting DPDK.

Signed-off-by: optimistyzy <optimistyzy@gmail.com>

### CODE REVIEW

File: src/os/bluestore/NVMEDevice.cc [hunk 1]
```diff
@@ -628,23 +628,51 @@ int NVMEManager::try_get(const string &sn_tag, SharedDriverData **driver)
...
+  char *master_core_arg;
 
   if (sn_tag.empty()) {
     r = -ENOENT;
     derr << __func__ << " empty serial number: " << cpp_strerror(r) << dendl;
     return r;
   }
 
+  core_value = strtoll(g_conf->bluestore_spdk_coremask.c_str(), NULL, 16);
+  for (int i = 0; i < 64; i++) {
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: "64" meaning?

Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: @yuyuyu101 I updated this code, it used to check the each bit in core_value.

Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: @yuyuyu101 I updated this code, it used to check the each bit in core_value.

Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: @yuyuyu101 I updated this code, it used to check the each bit in core_value.

Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: @yuyuyu101 I updated this code, it used to check the each bit in core_value.

**Comment by optimistyzy** [comment]
@yuyuyu101 I updated this code, it used to check each bit in core_value.

**Comment by tangwenjun3** [comment]
this patch tested  and worked well in my ceph，and solve the issues ( http://tracker.ceph.com/issues/19093)
succeed to deploy multiple SPDK blueStore OSD instances per node。

my osd config is that：
[osd.0]
    host = 127.0.0.1
    osd data = /data/osd/osd.0
    bluestore_block_path = spdk:55cd2e404c7e1063
    bluestore_spdk_socket_mem = 512,512
    bluestore_spdk_coremask = 0x3    #0、1  core


[osd.1]
    host = 127.0.0.1
    osd data = /data/osd/osd.1
    bluestore_block_path = spdk:55cd2e404bde228b
    bluestore_spdk_socket_mem = 512,512
    bluestore_spdk_coremask = 0XC00000  # 22、23 core

File: src/os/bluestore/NVMEDevice.cc [hunk 2]
```diff
@@ -628,23 +628,51 @@ int NVMEManager::try_get(const string &sn_tag, SharedDriverData **driver)
...
+      if(m_core < 0) {
+        m_core = i;
+      }
+    }
+  }
+
+  // at least two cores are needed for using spdk
+  if (core_num < 2) {
+    r = -ENOENT;
+    derr << __func__ << " invalid spdk coremask: " << cpp_strerror(r) << dendl;
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: suggest explicitly tell the reason like at least two cores needed

Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: updated.

**Comment by optimistyzy** [comment]
updated with suggested comments. @yuyuyu101

**Pull request merged**